File: src/main/java/com/example/bongabdodemo/utils/BongabdoCalculation.java
package com.example.bongabdodemo.utils;

import com.hellohasan.bongabdo.api.Bongabdo;
import com.hellohasan.bongabdo.api.BongabdoMethod;
import com.hellohasan.bongabdo.localization.BengaliLocalizationConfig;
import com.hellohasan.bongabdo.localization.BongabdoLocalizationConfig;
import com.hellohasan.bongabdo.localization.EnglishLocalizationConfig;
import org.springframework.context.i18n.LocaleContextHolder;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

public class BongabdoCalculation {

    private static final String DD_MMM_YYYY = "yyyy-MM-dd";

    /**
     * There are two available calculation methods: BANGLA_ACADEMY and INDIAN_DRIK_SIDDHANTA available in our library.
     * But you can extend the Bongabdo class and implement your own logic inside getBongabdoData method.
     * Example: SurjaShiddhantaBongabdo
     */
    public static String getBanglaAcademyBongabdo(Calendar calendar) {
        Bongabdo bongabdo = Bongabdo.getInstance(BongabdoMethod.BANGLA_ACADEMY);
        bongabdo.setLocalizationConfig(getLocalizationConfig());
        var bongabdoDate = bongabdo.getBongabdoData(
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
        );
        return bongabdoDate.getFullDate();
    }

    public static String getDrikShiddhantaBongabdo(Calendar calendar) {
        Bongabdo bongabdo = Bongabdo.getInstance(BongabdoMethod.INDIAN_DRIK_SIDDHANTA);
        bongabdo.setLocalizationConfig(getLocalizationConfig());
        var bongabdoDate = bongabdo.getBongabdoData(
                calendar.get(Calendar.YEAR),
                calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)
        );
        return bongabdoDate.getFullDate();
    }

    /**
     * You can implement your own localization config class for Hindi, Arabic or any other language.
     * Just create a subclass of BongabdoLocalizationConfig. Then return it from this method.
     * Example: HindiLocalizationConfig
     */
    private static BongabdoLocalizationConfig getLocalizationConfig() {
        Locale locale = LocaleContextHolder.getLocale();
        if ("bn".equals(locale.getLanguage())) {
            return new BengaliLocalizationConfig();
        } else {
            return new EnglishLocalizationConfig();
        }
    }

    public static String getDateString(Calendar calendar) {
        SimpleDateFormat formatter = new SimpleDateFormat(DD_MMM_YYYY, Locale.getDefault());
        return formatter.format(calendar.getTime());
    }

    public static Calendar getCalendarFromString(String dateString) {
        SimpleDateFormat formatter = new SimpleDateFormat(DD_MMM_YYYY, Locale.getDefault());
        try {
            Calendar calendar = Calendar.getInstance();
            calendar.setTime(formatter.parse(dateString));
            return calendar;
        } catch (ParseException e) {
            throw new IllegalArgumentException("Invalid date format: " + dateString);
        }
    }

}

File: src/main/kotlin/com/example/bongabdodemo/utils/BongabdoCalculation.kt
package com.example.bongabdodemo.utils

import com.hellohasan.bongabdo.api.Bongabdo
import com.hellohasan.bongabdo.api.BongabdoMethod
import com.hellohasan.bongabdo.localization.BengaliLocalizationConfig
import com.hellohasan.bongabdo.localization.BongabdoLocalizationConfig
import com.hellohasan.bongabdo.localization.EnglishLocalizationConfig
import org.springframework.context.i18n.LocaleContextHolder
import java.text.SimpleDateFormat
import java.util.*

private const val DD_MMM_YYYY = "yyyy-MM-dd"

object BongabdoCalculation {

    /**
     * There are two available calculation methods: [BongabdoMethod.BANGLA_ACADEMY] and [BongabdoMethod.INDIAN_DRIK_SIDDHANTA] available in our library.
     * But you can extend [Bongabdo] class and implement your own logic inside [getBongabdoData] method.
     * Example: [SurjaShiddhantaBongabdo]
     */
    fun getBanglaAcademyBongabdo(calendar: Calendar): String {
        val bongabdo = Bongabdo.getInstance(BongabdoMethod.BANGLA_ACADEMY)
        bongabdo.localizationConfig = getLocalizationConfig()
        val bongabdoDate = bongabdo.getBongabdoData(
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )
        return bongabdoDate.getFullDate()
    }

    fun getDrikShiddhantaBongabdo(calendar: Calendar): String {
        val bongabdo = Bongabdo.getInstance(BongabdoMethod.INDIAN_DRIK_SIDDHANTA)
        bongabdo.localizationConfig = getLocalizationConfig()
        val bongabdoDate = bongabdo.getBongabdoData(
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        )
        return bongabdoDate.getFullDate()
    }

    /**
     * You can implement your own localisation config class for Hindi, Arabic or any other language.
     * Just create a subclass of [BongabdoLocalizationConfig]. Then return it from this method.
     * Example: [HindiLocalisationConfig]
     * */
    private fun getLocalizationConfig(): BongabdoLocalizationConfig {
        val locale = LocaleContextHolder.getLocale()
        return if (locale.language == "bn")
            BengaliLocalizationConfig()
        else EnglishLocalizationConfig()
    }

    fun getDateString(calendar: Calendar): String {
        val formatter = SimpleDateFormat(DD_MMM_YYYY, Locale.getDefault())
        return formatter.format(calendar.time)
    }

    fun getCalendarFromString(dateString: String): Calendar {
        val formatter = SimpleDateFormat(DD_MMM_YYYY, Locale.getDefault())
        val date = formatter.parse(dateString)
        return Calendar.getInstance().apply {
            time = date ?: throw IllegalArgumentException("Invalid date format: $dateString")
        }
    }

}